crates/neon-runtime/src/napi/fun.rs

```diff
@@ -22,6 +23,77 @@ pub unsafe fn new(out: &mut Local, env: Env, callback: CCallback) -> bool {
     status == napi::Status::Ok
 }
 
+pub unsafe fn boxed<F>(env: Env, name: &str, callback: F) -> Result<Local, napi::Status>
+where
+    F: FnMut(Env, napi::CallbackInfo) -> Local + Send + 'static,
+{
+    let mut out = MaybeUninit::uninit();
+    let data = Box::into_raw(Box::new(callback));
+    let status = napi::create_function(
+        env,
+        name.as_ptr().cast(),
+        name.len(),
+        Some(call_boxed::<F>),
+        data.cast(),
+        out.as_mut_ptr(),
+    );
+
+    if status == napi::Status::PendingException {
+        Box::from_raw(data);
+
+        return Err(status);
+    }
+
+    assert_eq!(status, napi::Status::Ok);
+
+    let out = out.assume_init();
+
+    #[cfg(feature = "napi-5")]
+    {
+        unsafe extern "C" fn drop_function<F>(
+            _env: Env,
+            finalize_data: *mut c_void,
+            _finalize_hint: *mut c_void,
+        ) {
+            Box::from_raw(finalize_data.cast::<F>());
+        }
+
+        let status = napi::add_finalizer(
+            env,
+            out,
+            data.cast(),
+            Some(drop_function::<F>),
+            ptr::null_mut(),
+            ptr::null_mut(),
+        );
+
+        assert_eq!(status, napi::Status::Ok);
```

Comment:
It's not safe to drop the boxed callback on failure here because the finalizer has already been given to Node. We do not have a guarantee that it won't be called.

Reply:
Is this a potential leak then?